File: sentinel/saas/schemas/audit.py
"""
Audit Log Schemas - Request/Response Models
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


# ============================================================================
# THREAT DETAILS SCHEMAS
# ============================================================================

class PIIEntityDetail(BaseModel):
    """Detailed PII entity information for display"""
    entity_type: str
    masked_value: Optional[str] = None  # Only for admin/owner users
    redaction_strategy: str
    start_position: int
    end_position: int
    confidence: float
    detection_method: str
    token_id: str


class InjectionDetail(BaseModel):
    """Injection attempt details"""
    injection_type: str
    confidence: float
    matched_patterns: List[str] = Field(default_factory=list)
    severity: str


class ContentViolationDetail(BaseModel):
    """Content policy violations"""
    violation_type: str
    matched_terms: List[str] = Field(default_factory=list)
    severity: str


class ThreatDetails(BaseModel):
    """Complete threat information structure"""
    pii: List[PIIEntityDetail] = Field(default_factory=list)
    injections: List[InjectionDetail] = Field(default_factory=list)
    content_violations: List[ContentViolationDetail] = Field(default_factory=list)
    total_threat_count: int = 0
    blocking_reasons: List[str] = Field(default_factory=list)


# ============================================================================
# RESPONSE SCHEMAS
# ============================================================================

class AuditLogResponse(BaseModel):
    """Audit log entry"""
    id: int
    timestamp: datetime
    session_id: str
    request_id: Optional[str]
    org_id: Optional[str]
    workspace_id: Optional[str]
    user_id: Optional[str]
    user_role: Optional[str]
    ip_address: Optional[str]
    user_input: str
    input_length: int
    blocked: bool
    risk_score: Optional[float]
    risk_level: Optional[str]
    pii_detected: bool
    pii_entities: Optional[list]
    redacted_count: int
    injection_detected: bool
    injection_type: Optional[str]
    injection_confidence: Optional[float]
    escalated: bool
    escalated_to: Optional[str]
    metadata: Optional[dict]
    threat_details: Optional[ThreatDetails] = None

    class Config:
        from_attributes = True


class AuditLogListResponse(BaseModel):
    """List of audit logs with pagination"""
    logs: list[AuditLogResponse]
    total: int
    page: int
    page_size: int

File: web/src/api/audit.ts
/**
 * Audit Logs API Client
 */

import { apiClient } from './client';

export interface AuditLog {
  id: number;
  timestamp: string;
  session_id: string;
  request_id?: string;
  org_id?: string;
  workspace_id?: string;
  user_id?: string;
  user_role?: string;
  ip_address?: string;
  user_input: string;
  input_length: number;
  blocked: boolean;
  risk_score?: number;
  risk_level?: string;
  pii_detected: boolean;
  pii_entities?: Record<string, any>;
  redacted_count: number;
  injection_detected: boolean;
  injection_type?: string;
  injection_confidence?: number;
  escalated: boolean;
  escalated_to?: string;
  metadata?: Record<string, any>;
}

export interface AuditLogListResponse {
  logs: AuditLog[];
  total: number;
  page: number;
  page_size: number;
}

export interface AuditLogFilters {
  start_date?: string;
  end_date?: string;
  workspace_id?: string;
  user_id?: string;
  blocked_only?: boolean;
  pii_only?: boolean;
  injection_only?: boolean;
  min_risk_score?: number;
  search?: string;
  page?: number;
  page_size?: number;
}

export const auditApi = {
  /**
   * List audit logs with filtering
   */
  listAuditLogs: async (filters: AuditLogFilters = {}): Promise<AuditLogListResponse> => {
    const params = new URLSearchParams();

    if (filters.start_date) params.append('start_date', filters.start_date);
    if (filters.end_date) params.append('end_date', filters.end_date);
    if (filters.workspace_id) params.append('workspace_id', filters.workspace_id);
    if (filters.user_id) params.append('user_id', filters.user_id);
    if (filters.blocked_only) params.append('blocked_only', 'true');
    if (filters.pii_only) params.append('pii_only', 'true');
    if (filters.injection_only) params.append('injection_only', 'true');
    if (filters.min_risk_score !== undefined) params.append('min_risk_score', filters.min_risk_score.toString());
    if (filters.search) params.append('search', filters.search);
    if (filters.page) params.append('page', filters.page.toString());
    if (filters.page_size) params.append('page_size', filters.page_size.toString());

    const response = await apiClient.get<AuditLogListResponse>(`/audit-logs?${params.toString()}`);
    return response.data;
  },

  /**
   * Get single audit log by ID
   */
  getAuditLog: async (logId: number): Promise<AuditLog> => {
    const response = await apiClient.get<AuditLog>(`/audit-logs/${logId}`);
    return response.data;
  },
};
